### PR TITLE
IAP brand & client

### PR DESCRIPTION
**Pull request by drebes**
Fixes terraform-providers/terraform-provider-google#1287

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_iap_brand`
```
```release-note:new-resource
`google_iap_client`
```

### CODE REVIEW

File: templates/terraform/examples/iap_identity_aware_proxy_client.tf.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  org_id     = "%{org_id}"
+}
+
+resource "google_project_service" "project_service" {
+  project = google_project.project.project_id
+  service = "iap.googleapis.com"
+}
+
+resource "google_iap_brand" "project_brand" {
+  support_email     = "support@example.com"
```
Comment (drebes) [inline templates/terraform/examples/iap_identity_aware_proxy_client.tf.erb]: This acceptance test won't succeed unless the specified group is owned by the service account running the test.

File: third_party/terraform/tests/resource_iap_brand_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  org_id     = "%{org_id}"
+}
+
+resource "google_project_service" "project_service" {
+  project = google_project.project.project_id
+  service = "iap.googleapis.com"
+}
+
+resource "google_iap_brand" "project_brand" {
+  support_email     = "support@example.com"
```
Comment (drebes) [inline third_party/terraform/tests/resource_iap_brand_test.go]: This acceptance test won't succeed unless the specified group is owned by the service account running the test.

**Comment by drebes** [comment]
Fixes terraform-providers/terraform-provider-google#1287

**Comment by danawillow** [comment]
drive-by: the "fixes" statement won't automatically close the issue if it's in a comment- it needs to be in the PR description or a commit (I usually do the PR description like in https://github.com/GoogleCloudPlatform/magic-modules/pull/3063)

File: products/iap/api.yaml [hunk 3]
```diff
@@ -135,3 +135,66 @@ objects:
...
+    base_url: 'projects/{{project}}/brands'
+    self_link: '{{name}}'
+    input: true
+    description: |
+      OAuth brand data.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Output only. Identifier of the brand. NOTE: GCP project number achieves
```
Comment (slevenick) [inline products/iap/api.yaml]: Can we document the format of `name`? In the generated documentation we have the import format listed as just `{name}` which will be difficult to figure out if we don't list the actual format of `name` somewhere

File: products/iap/api.yaml [hunk 4]
```diff
@@ -135,3 +135,66 @@ objects:
...
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Output only. Unique identifier of the OAuth client.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'brand'
+        description: |
+          Identifier of the brand to which this `IdentityAwareProxyClients``
```
Comment (slevenick) [inline products/iap/api.yaml]: Can we document the format of `brand` here? We show the import id of this resource as `{brand}/identityAwareProxyClients/{client_id}` but a user may not know the format of `brand`.

File: third_party/terraform/tests/resource_iap_brand_test.go (hunk 2 above)
Comment (drebes) [reply third_party/terraform/tests/resource_iap_brand_test.go]: Service accounts cannot be specified. Users can be created but not with the Google provider, only the 3rd-party G Suite one.

**Comment by pvillard31** [comment]
_Sorry for the comment/question, feel free to redirect me to the right place if I should ask this somewhere else._

If I correctly understand this PR, its purpose is to provide a way to create the "APIs & Services / OAuth consent screen". However, I do not see how to specify information such as authorised domains, homepage link, etc. Am I missing something?

Also I see that client ID and client secret are made available but how to specify  "Authorised JavaScript origins" and "Authorised redirect URIs". Is it out of scope of this PR?

File: third_party/terraform/tests/resource_iap_brand_test.go (hunk 2 above)
Comment (drebes) [reply third_party/terraform/tests/resource_iap_brand_test.go]: I've added more details to the supportEmail description to make it clear what are the valid options.

File: products/iap/api.yaml [hunk 5]
```diff
@@ -135,3 +135,66 @@ objects:
...
+        description: |
+          Application name displayed on OAuth consent screen.
+        required: true
+      - !ruby/object:Api::Type::Boolean
+        name: 'orgInternalOnly'
+        description: |
+          Whether the brand is only intended for usage inside the GSuite organization only.
+        output: true
+  - !ruby/object:Api::Resource
+    name: 'IdentityAwareProxyClient'
```
[the remark this answers is not in the document]
Comment (drebes) [reply products/iap/api.yaml]: Indeed looks much better. Changed.

File: products/iap/api.yaml [hunk 6]
```diff
@@ -135,3 +135,66 @@ objects:
         name: 'name'
         description: Name of the instance.
         required: true
+  - !ruby/object:Api::Resource
+    name: 'Brand'
+    base_url: 'projects/{{project}}/brands'
+    self_link: '{{name}}'
+    input: true
+    description: |
+      OAuth brand data.
```
[the remark this answers is not in the document]
Comment (drebes) [reply products/iap/api.yaml]: Done.

File: templates/terraform/extra_schema_entry/iap_identity_aware_proxy_client.erb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+"client_id": {
```
[the remark this answers is not in the document]
Comment (drebes) [reply templates/terraform/extra_schema_entry/iap_identity_aware_proxy_client.erb]: Done. post_create is still needed as the name/clientId is not returned on the create call, though.

File: products/iap/api.yaml (hunk 3 above)
Comment (drebes) [reply products/iap/api.yaml]: Done.

File: products/iap/api.yaml (hunk 4 above)
Comment (drebes) [reply products/iap/api.yaml]: Done.

**Comment by slevenick** [comment]
https://github.com/GoogleCloudPlatform/magic-modules/pull/3162 Should fix the issues I ran into around the beta provider and base path names.

I'm happy with the current implementation, I would just like to get integration tests working for this so that we can run them via CI.

I think all we need is to add a new environment variable to `provider_test.go` that will represent the organization domain. Then we can depend on `support@{org-domain}` existing and the service account running the tests to be an admin for that group email (I'll go set this up manually)

**Comment by slevenick** [comment]
Hey @drebes do you think you will have a chance to add support for those tests via the environment variable sometime soon?

I can also pick this up and add the changes to your branch, just let me know if you want me to do that

**Comment by drebes** [comment]
Sure I can add them (sorry, I didn't understand "we need is to add a new environment variable" to mean me, but I can look into it.)

**Comment by drebes** [comment]
Added the configurable domain for the acceptance tests.

You need to create a group "support@<acceptance_test_domain>" and make the acceptance test service account an owner of such group, and then set `GOOGLE_ORG_DOMAIN` as <acceptance_test_domain>.

**Comment by slevenick** [comment]
Awesome, looks good! I got it to pass locally, so I'll modify the CI test org to add this. Can you rebase this from master, looks like a couple things are out of sync

**Comment by drebes** [comment]
There you go. Rebased with the collapsed sidebar changes that were causing the conflict. Which also made me realized that I hadn't updated the sidebar for `iap_client` since I had updated it to the short form. Should be good now.

**Pull request merged**